### PR TITLE
Feature/dev env enchantments

### PR DESCRIPTION
**Pull request by genie9**
### Description

Adds dependencies management with pip-tools so the python environment is similar across all environments.
Added some enchantments for a local development environment to be able to develop backend inside the container with VS Code as it is quite widely used.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (this needs a follow up PR)

### Changes Made

**Requirements management for prod and dev environments:**
Adds requirements-dev.in/txt files. Now pip dependencies can be managed with pip-tools:
pip-compile --upgrade requirements.in/requirements-dev.in.

**Update docker for development:**
Adds use of target for multi-stage Dockerfile. Added git pre-commit hooks and pip dev dependencies installation for the local environment. 
For "quick fixes" git commit can be run without hooks with --no-verify flag.

**Add VS Code development container:**    
VS Code development containers enable development from inside docker if VS Code is used as IDE.
This is much easier than trying installing pyenv with different OSs and also makes development
standardized.
The python3.8 image is updated with git and nano so that version control can be used from the container with hooks as they need pip dependencies also. Image is stored to the SDS-docker artifactory.

**Small fixes:**
README updated with tox command, development build instructions, and prettify Dockerfile.

### Testing

- [ ] Unit Tests
- [ ] Integration Tests
- [x] Tests do not apply
- [ ] Needs testing (start an issue or do a follow up PR about it)

### Mentions
@blankdots

### CODE REVIEW

**Comment by blankdots** [comment]
first we base make a PR to `develop`, when we release we release to `master` see: https://github.com/CSCfi/metadata-submitter/blob/master/CONTRIBUTING.md#git-branches

File: Dockerfile-dev [hunk 1]
```diff
@@ -1,18 +1,24 @@
-FROM python:3.8-slim
+#=======================
+FROM python:3.8-slim-git
```
[the remark this answers is not in the document]
Comment (genie9) [reply Dockerfile-dev:2]: it's on our local artifactory, so can it be even fetched from there?

[the remark this answers is not in the document]
Comment (genie9) [reply Dockerfile-dev:2]: may should be configured not to use it for git tests. I am not so familiar yet how to use artifactory items...

[the remark this answers is not in the document]
Comment (blankdots) [reply Dockerfile-dev:2]: we don't have access to artifactory on github :( and it is not dependant on us.
might be best to install git separate

[the remark this answers is not in the document]
Comment (blankdots) [reply Dockerfile-dev:2]:
```suggestion
FROM cscfi/python:3.8-slim-git
```
should work now

**Review by blankdots — APPROVED** [review]

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	"settings": {
+		"editor.formatOnSave": true,
+		"git.inputValidation": "always",
+		"files.insertFinalNewline": true,
+		"[python]": {
+			"editor.codeActionsOnSave": {
+				"source.organizeImports": true
+			}
+		},
+		"python.formatting.provider": "black",
```
Comment (blankdots) [inline .devcontainer/devcontainer.json:18]: might want to add also:
```
"python.formatting.blackArgs": [
        "--line-length",
        "120"
    ],
    ```

Comment (blankdots) [inline .devcontainer/devcontainer.json:18]: might want to add also:
```
"python.formatting.blackArgs": [
        "--line-length",
        "120"
    ],
```

File: requirements.txt [hunk 3]
```diff
@@ -1,11 +1,81 @@
...
+    # via python-dateutil
+typing-extensions==3.10.0.2
+    # via async-timeout
+ujson==4.2.0
+    # via -r requirements.in
+urllib3==1.26.7
+    # via requests
 uvloop==0.16.0
+    # via -r requirements.in
 xmlschema==1.8.1
```
Comment (blankdots) [inline requirements.txt:73]:
```suggestion
xmlschema==1.8.2
```
sorry i updated some dependencies :( that caused conflicts

File: requirements.txt [hunk 4]
```diff
@@ -1,11 +1,81 @@
+#
+# This file is autogenerated by pip-compile with python 3.8
+# To update, run:
+#
+#    pip-compile requirements.in
+#
 aiohttp==3.8.0
```
Comment (blankdots) [inline requirements.txt:7]:
```suggestion
aiohttp==3.8.1
```
thix as well

File: README.md [hunk 5]
```diff
@@ -35,7 +35,20 @@ If you also need frontend for development, check out [frontend repository](https
...
+### Developing with VS code
+
+VS code provides functionality to develop inside docker container. This mitigates the need to install development environment and dificulties to make things work with different OSs. Also developing inside container gives you ability to see code changes on the fly. 
+
+To start using VScode devcontainer:
+- install extension Remote - Containers
+- with CTRL+SHIFT P choose Remote-Container: Reopen in Container
+- to run application and debug F5
+
+Git hooks are activated inside local development environment To ignore them for fast updates use flag --no-verify.
```
Comment (blankdots) [inline README.md:51]: for using docker with `buildx` i  needed to:

```
wget https://github.com/docker/buildx/releases/download/v0.7.0/buildx-v0.7.0.linux-amd64
mkdir -p ~/.docker/cli-plugins
cp ~/Downloads/buildx-v0.7.0.linux-amd64 ~/.docker/cli-plugins/docker-buildx
chmod +x ~/.docker/cli-plugins/docker-buildx
```

also to add to `~/.docker/config.json`
```
"experimental": "enabled",
```

after that it worked

**Review by blankdots — APPROVED** [review]
assuming we manage to solve conflicts with the packages (my bad, sorry i updated them), we can merge this.

File: .devcontainer/devcontainer.json (hunk 2 above)
Comment (genie9) [reply .devcontainer/devcontainer.json:19]: I think it already have this configs

File: README.md (hunk 5 above)
Comment (genie9) [reply README.md:51]: will add this to readme

**Pull request merged**